### PR TITLE
Adds project list descriptions generation

### PR DESCRIPTION
**Pull request by dominoFire**
## Bug

Fixes: https://github.com/NuGet/Client.Engineering/issues/119
Regression: No
* Last working version: N/A
* How are we preventing it in future: 

## Fix

Adds a MSBuild task to generate a markdown file with the project descriptions

## Testing/Validation

Tests Added: No  
Reason for not adding tests: Engineering infra.
Validation: Manual run

### CODE REVIEW

**Review by zkat — APPROVED** [review]
yay! This is great!

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -93,6 +93,7 @@ NuGet members may contribute directly to the main remote.
 
 - `$(NuGetClientRoot)\artifacts\VS15` - this folder will contain the Package Manager extension (`NuGet.Tools.vsix`) and NuGet command-line client application (`NuGet.exe`)
 - `$(NuGetClientRoot)\artifacts\nupkgs` - this folder will contain all our projects packages
+- `$(NuGetClientRoot)\artifacts\docs-generated` - Containes generated docs. Created with `msbuild .\build\build.proj /t:GenerateDocs`
```
Comment (nkolev92) [inline CONTRIBUTING.md]: I think the docs should be generated into the docs folder at the repo root. 

Is there a good reason why we'd an intermediate doc?

File: build/build.proj [hunk 2]
```diff
@@ -415,4 +415,28 @@
...
+  <!--
+    ============================================================
+    Generate documentation for source projects
+    ============================================================
+  -->
+  <UsingTask AssemblyFile="$(RepositoryRootDirectory)\tools-local\doc.tasks\bin\Debug\netstandard2.0\doc.tasks.dll" TaskName="NuGetTasks.GenerateMarkdownDoc" />
+  <Target Name="GenerateDocs">
+    <MsBuild
+      Projects="$(RepositoryRootDirectory)\tools-local\doc.tasks\doc.tasks.csproj"
+      Targets="Restore;Build"
```
Comment (nkolev92) [inline build/build.proj]: Don't do this. 
See: https://docs.microsoft.com/en-us/nuget/reference/msbuild-targets#restoring-and-building-with-one-msbuild-command.

File: tools-local/doc.tasks/GenerateMarkdownDoc.cs [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+
+        public override bool Execute()
+        {
+            var outputFilePath = OutputFile.GetMetadata("FullPath");
+
+            var dirname = Path.GetDirectoryName(outputFilePath);
+            Directory.CreateDirectory(dirname);
+            Log.LogMessage(MessageImportance.Low, $"Created directories: {dirname}");
+
+            Log.LogMessage(MessageImportance.Low, $"Creating documentation: {outputFilePath}");
```
Comment (nkolev92) [inline tools-local/doc.tasks/GenerateMarkdownDoc.cs:48]: Maybe it's a good idea to write a basically intro to the overview and then have 2 categories of projects in the list. 
Specifically have NuGet.Clients and NuGet.Core.
But that can be done later. 
We can make incremental changes.

File: tools-local/doc.tasks/GenerateMarkdownDoc.cs [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+            using (var file = new StreamWriter(outputFilePath))
+            {
+                file.WriteLine("---");
+                file.WriteLine($"date-generated: {DateTime.Now:s}");
+                file.WriteLine("---\n");
+
+                file.WriteLine($"# {Title}\n\n");
+
+                for (int i = 0; i < ProjectFiles.Length; i++)
```
Comment (nkolev92) [inline tools-local/doc.tasks/GenerateMarkdownDoc.cs]: Sort the ProjectFiles by name.

**Review by nkolev92 — COMMENTED** [review]
Please move the generated doc to docs/*

It'll be great if you can add a sample of the generated doc in this PR.

File: build/build.proj (hunk 2 above)
Comment (dominoFire) [reply build/build.proj]: Changed to two MSBuild task invocations

File: CONTRIBUTING.md (hunk 1 above)
Comment (dominoFire) [reply CONTRIBUTING.md]: Code-generated artifacts normally are not checked-in in a source control repo. I will add the generated artifact.

File: tools-local/doc.tasks/GenerateMarkdownDoc.cs (hunk 4 above)
Comment (dominoFire) [reply tools-local/doc.tasks/GenerateMarkdownDoc.cs]: Sorted.

File: tools-local/doc.tasks/GenerateMarkdownDoc.cs (hunk 3 above)
Comment (dominoFire) [reply tools-local/doc.tasks/GenerateMarkdownDoc.cs]: Added tests projects to the overview.

Comment (dominoFire) [reply tools-local/doc.tasks/GenerateMarkdownDoc.cs:48]: Added test projects to the list.

File: CONTRIBUTING.md (hunk 1 above)
Comment (nkolev92) [reply CONTRIBUTING.md]: I don't get what you mean.
The doc that this tool is supposed to generate can be a part of the PR. 
It's not a "code-generated" artifact comparable to dlls etc.

**Comment by nkolev92** [comment]
@dominoFire Please re-request review when ready. 

The status was a bit ambiguous to me based on your last comment :)

File: CONTRIBUTING.md (hunk 1 above)
Comment (dominoFire) [reply CONTRIBUTING.md]: Moved the project-list.md to docs\ folder

File: tools-local/doc.tasks/GenerateMarkdownDoc.cs (hunk 3 above)
Comment (dominoFire) [reply tools-local/doc.tasks/GenerateMarkdownDoc.cs:48]: Also, I added a brief description in project-list.md

File: docs/project-list.md [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+
+
+
+## Projects
+
+All shipped NuGet libraries and clients lives in `src/` folder.
+
+Projects count: 37
+
+- [`src\NuGet.Clients\NuGet.CommandLine\NuGet.CommandLine.csproj`](../src/NuGet.Clients/NuGet.CommandLine/NuGet.CommandLine.csproj): NuGet Command Line Interface.
```
Comment (nkolev92) [inline docs/project-list.md]: How about we use the project name as the link name? 
Right now it looks kind of clunky.

File: docs/project-list.md [hunk 6]
```diff
@@ -0,0 +1,103 @@
+---
+date-generated: 2020-03-24T14:44:10
+tool: NuGetTasks.GenerateMarkdownDoc
+---
+
+
+# NuGet Project Overview
+
+Below is a list of all source code projects for NuGet libraries and supported NuGet clients
+
```
Comment (nkolev92) [inline docs/project-list.md]: Lots of unnecessary empty lines here :)

File: docs/project-list.md [hunk 7]
```diff
@@ -0,0 +1,103 @@
+---
+date-generated: 2020-03-24T14:44:10
+tool: NuGetTasks.GenerateMarkdownDoc
+---
+
+
+# NuGet Project Overview
```
Comment (nkolev92) [inline docs/project-list.md]: I'd also call the file "project-overview.md"

File: tools-local/doc.tasks/GenerateMarkdownDoc.cs [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+            Log.LogMessage(MessageImportance.Low, $"Creating documentation: {outputFilePath}");
+            using (var file = new StreamWriter(outputFilePath))
+            {
+                file.WriteLine("---");
+                file.WriteLine($"date-generated: {DateTime.Now:s}");
+                file.WriteLine($"tool: {typeof(GenerateMarkdownDoc).FullName}");
+                file.WriteLine("---");
+
+                file.WriteLine($"\n\n# {Title}\n");
+                file.WriteLine("Below is a list of all source code projects for NuGet libraries and supported NuGet clients\n");
```
Comment (nkolev92) [inline tools-local/doc.tasks/GenerateMarkdownDoc.cs]: This feels redundant.

File: tools-local/doc.tasks/GenerateMarkdownDoc.cs [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+
+                file.WriteLine($"\n\n## {ProjectsSubtitle}\n");
+                file.WriteLine("All shipped NuGet libraries and clients lives in `src/` folder.\n");
+                file.WriteLine($"Projects count: {ProductProjects.Length}\n");
+                SortAndWriteDescriptions(ProductProjects, file);
+
+                file.WriteLine($"\n\n## {TestProjectsSubtitle}\n");
+                file.WriteLine("Most production assemblies has an associated test project, whose name ends with `.Test`.\n");
+                file.WriteLine($"Test Projects count: {TestProjects.Length}\n");
+                SortAndWriteDescriptions(TestProjects, file);
```
Comment (heng-liu) [inline tools-local/doc.tasks/GenerateMarkdownDoc.cs]: Shall we have a separate test-overview.md file for test projects?

File: docs/project-list.md (hunk 7 above)
Comment (dominoFire) [reply docs/project-list.md]: Changed to project-overview.md

File: tools-local/doc.tasks/GenerateMarkdownDoc.cs (hunk 9 above)
Comment (dominoFire) [reply tools-local/doc.tasks/GenerateMarkdownDoc.cs]: We can have a separte file. Now, the task supports generating two tasks. Please, review again.

File: tools-local/doc.tasks/GenerateMarkdownDoc.cs (hunk 8 above)
Comment (dominoFire) [reply tools-local/doc.tasks/GenerateMarkdownDoc.cs]: Reprhased summary

File: docs/project-list.md (hunk 6 above)
Comment (dominoFire) [reply docs/project-list.md]: Removed some empty lines. Added double empty lines between sections. Do you want exactly one empty line between sections ?

File: docs/project-list.md (hunk 5 above)
Comment (dominoFire) [reply docs/project-list.md]: Changed to project name.

File: tools-local/doc.tasks/GenerateMarkdownDoc.cs (hunk 9 above)
Comment (nkolev92) [reply tools-local/doc.tasks/GenerateMarkdownDoc.cs]: cc @donnie-msft thoughts on this. 
https://github.com/NuGet/Client.Engineering/issues/239 will be impacted by that decision.

File: docs/project-overview.md [hunk 10]
```diff
@@ -0,0 +1,96 @@
+---
```
Comment (nkolev92) [inline docs/project-overview.md:1]: You are missing the func tests.

File: docs/project-overview.md [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+date-generated: 2020-03-30T13:41:16
+tool: NuGetTasks.GenerateMarkdownDoc
+---
+
+
+# NuGet Project Overview
+
+Below is a list of projects contained in the NuGet.Client repo, organized by source and test projects
```
Comment (nkolev92) [inline docs/project-overview.md]: Use punctuation at the end.

Comment (nkolev92) [reply docs/project-overview.md]: > source and test projects

below you call them product projects...i think that's a better name, so maybe use it here if you agree?

**Review by nkolev92 — APPROVED** [review]
Looks good, however, you are the missing the functional tests.

Approving now...just make sure you've included all projects.

File: tools-local/doc.tasks/GenerateMarkdownDoc.cs (hunk 9 above)
Comment (heng-liu) [reply tools-local/doc.tasks/GenerateMarkdownDoc.cs]: As we discussed offline, we can have a separate test-overview.md file.

File: docs/project-overview.md (hunk 11 above)
Comment (dominoFire) [reply docs/project-overview.md]: Agree. Changed to `...by product and test projects.`.

File: docs/project-overview.md (hunk 10 above)
Comment (dominoFire) [reply docs/project-overview.md:1]: 😅 Added functional tests section.

File: tools-local/doc.tasks/GenerateMarkdownDoc.cs (hunk 9 above)
Comment (donnie-msft) [reply tools-local/doc.tasks/GenerateMarkdownDoc.cs]: What is the rationale for having separate files?

**Pull request merged**